app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/OnlineLibraryFragment.kt

```diff
@@ -301,30 +306,59 @@ class OnlineLibraryFragment : BaseFragment(), FragmentActivityExtensions {
     }
   }
 
+  private fun requestExternalStoragePermission() {
+    ActivityCompat.requestPermissions(
+      requireActivity(), arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE),
+      REQUEST_STORAGE_PERMISSION
+    )
+  }
+
+  override fun onRequestPermissionsResult(
+    requestCode: Int,
+    permissions: Array<out String>,
+    grantResults: IntArray
+  ) {
+    super.onRequestPermissionsResult(requestCode, permissions, grantResults)
+    if (requestCode == REQUEST_STORAGE_PERMISSION &&
+      grantResults.isNotEmpty() &&
+      grantResults[0] != PERMISSION_GRANTED
+    ) {
+      context.toast(R.string.request_write_storage)
+      requestExternalStoragePermission()
+    }
+  }
+
+  private fun hasPermission(permission: String): Boolean =
+    ContextCompat.checkSelfPermission(requireActivity(), permission) == PERMISSION_GRANTED
+
   private fun onBookItemClick(item: LibraryListItem.BookItem) {
-    when {
-      isNotConnected -> {
-        noInternetSnackbar()
-        return
-      }
-      noWifiWithWifiOnlyPreferenceSet -> {
-        dialogShower.show(WifiOnly, {
-          sharedPreferenceUtil.putPrefWifiOnly(false)
-          downloadFile(item.book)
-        })
-        return
+    if (hasPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
```

Comment:
We should check here if we have `checkExternalStorageWritePermission()`, if we don't have permission and the user has already denied permission more than once then we'll ask rationally like this
```kt
private fun checkExternalStorageWritePermission(): Boolean { 
    return hasPermission(WRITE_EXTERNAL_STORAGE).also { permissionGranted ->
      if (!permissionGranted) {
        if (shouldShowRationale(WRITE_EXTERNAL_STORAGE)) {
          alertDialogShower.show(
            KiwixDialog.StoragePermissionRationale,
            ::requestExternalStoragePermission
          )
        } else {
          requestExternalStoragePermission()
        }
      }
    }
  }

 private fun shouldShowRationale(writeExternalStorage: String) =
    ActivityCompat.shouldShowRequestPermissionRationale(requireActivity(), writeExternalStorage)
```

Reply:
done